### PR TITLE
windows_workgroup Resource for joining Windows Workgroups

### PR DESCRIPTION
**Pull request by derekgroh**
### Description

Create resource to handle joining windows workstation to workgroup

### Issues Resolved

Slack discussion requesting the feature.

### Check List

- [X] New functionality includes tests
- [X] All tests pass
- [N/A] RELEASE\_NOTES.md has been updated if required (not required for bugfixes, required for API changes)
- [X] All commits have been signed-off for the Developer Certificate of Origin. See <https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco>

### CODE REVIEW

**Comment by derekgroh** [comment]
Hello,
Believe I'm missing something with getting the tests to pass, not sure if it is an environment issue with my workstation or if I overlooked something.
```PS D:\cookbooks\derekgroh\chef> chef exec rspec D:\cookbooks\derekgroh\chef\spec\unit\resource\windows_workgroup_spec.rb
C:/opscode/chefdk/embedded/lib/ruby/gems/2.5.0/gems/rspec-support-3.7.1/lib/rspec/support.rb:124:in `warn' : WARNING: Shared example group 'with a chef repo' has been previously defined at:
  C:/opscode/chefdk/embedded/lib/ruby/gems/2.5.0/gems/cheffish-14.0.1/lib/cheffish/rspec/repository_support.rb:12
...and you are now defining it at:
  D:/cookbooks/derekgroh/chef/spec/support/shared/integration/integration_helper.rb:111
The new definition will overwrite the original one.
 (StructuredWarnings::BuiltInWarning)
An error occurred while loading ./spec/unit/resource/windows_workgroup_spec.rb.
Failure/Error:
  describe Chef::Resource::WindowsWorkgroup do
    let(:resource) { Chef::Resource::WindowsWorkgroup.new("example") }

    it "sets resource name as :windows_workgroup" do
      expect(resource.resource_name).to eql(:windows_workgroup)
    end

    it "the workgroup_name property is the name_property" do
      expect(resource.workgroup_name).to eql("example")
    end

NameError:
  uninitialized constant Chef::Resource::WindowsWorkgroup
  Did you mean?  Chef::Resource::WindowsShortcut
# ./spec/unit/resource/windows_workgroup_spec.rb:20:in `<top (required)>'
Run options:
  include {:focus=>true}
  exclude {:provider=>#<Proc:./spec/spec_helper.rb:211>, :arch=>#<Proc:./spec/spec_helper.rb:205>, :ruby=>"2.5.1", :chef=>"14.4.17", :not_intel_64bit=>true, :rhel7=>true, :rhel6=>true, :rhel5=>true, :rhel=>true, :not_wpar=>true, :broken=>true, :openssl_lt_101=>true, :requires_root=>true, :selinux_only=>true, :debian_family_only=>true, :aix_only=>true, :linux_only=>true, :unix_only=>true, :solaris_only=>true, :windows_service_requires_assign_token=>true, :windows_not_domain_joined_only=>true, :windows_powershell_dsc_only=>true, :ruby32_only=>true, :windows_lt_10=>true, :windows_nano_only=>true, :windows32_only=>true, :windows_2008r2_or_later=>true, :win2012r2_only=>true, :mac_osx_only=>true, :appveyor_only=>true, :volatile_from_verify=>false, :volatile=>true, :external=>true}

All examples were filtered out; ignoring {:focus=>true}

All examples were filtered out

Finished in 0.00163 seconds (files took 40.76 seconds to load)
0 examples, 0 failures, 1 error occurred outside of examples```

**Comment by tas50** [comment]
@derekgroh Almost there! You'll want to wire up the resource to load by requiring it here https://github.com/chef/chef/blob/master/lib/chef/resources.rb

**Comment by tas50** [comment]
Pinging @stuartpreston for thoughts.

File: lib/chef/resource/windows_workgroup.rb [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+               required: true
+
+      property :password, String,
+               description: "The password for the local user.",
+               required: true
+
+      property :reboot, Symbol,
+               equal_to: [:immediate, :delayed, :never, :request_reboot, :reboot_now],
+               validation_message: "The reboot property accepts :immediate (reboot as soon as the resource completes), :delayed (reboot once the Chef run completes), and :never (Don't reboot)",
+               description: "Controls the system reboot behavior post domain joining. Reboot immediately, after the Chef run completes, or never. Note that a reboot is necessary for changes to take effect.",
```
Comment (tas50) [inline lib/chef/resource/windows_workgroup.rb]: This description still mentions the domain

**Comment by derekgroh** [comment]
@stuartpreston Good catch

1. This might be the better implementation given 3.
2. Command fails rerunning the resource, added a check
3. The powershell command can be run without credentials, but may error out in powershell due to lack of permissions, not sure the best way to handle this case. Trying to work on some test cases.

**Comment by stuartpreston** [comment]
Hi @derekgroh thanks for this I'm putting it in my queue for review now.

**Comment by derekgroh** [comment]
@stuartpreston Awesome, let me know if something was overlooked, looking to learn.

File: lib/chef/resource/windows_workgroup.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+class Chef
+  class Resource
+    class WindowsWorkgroup < Chef::Resource
+      resource_name :windows_workgroup
+      provides :windows_workgroup
+
+      include Chef::Mixin::PowershellOut
+
+      description "Use the windows_workgroup resource to join change the workgroup of a machine."
+      introduced "14.4"
```
Comment (tas50) [inline lib/chef/resource/windows_workgroup.rb]: This will need to be updated to 14.5 now.

**Comment by lamont-granquist** [comment]
i'm good with this, but can't really review the windows code...

**Comment by stuartpreston** [comment]
Thanks - I haven't forgotten about this but other work has taken priority so far this week.

File: lib/chef/resource/windows_workgroup.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            :reboot_now
+          when :delayed
+            :request_reboot
+          else
+            reboot_action
+          end
+        end
+
+        def workgroup_member?
+          node_workgroup = powershell_out!("Get-WmiObject -Class Win32_ComputerSystem).Workgroup")
```
Comment (stuartpreston) [inline lib/chef/resource/windows_workgroup.rb]: This is missing an opening bracket - should be
```
node_workgroup = powershell_out!("(Get-WmiObject -Class Win32_ComputerSystem).Workgroup")
```

File: lib/chef/resource/windows_workgroup.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+      # define this again so we can default it to true. Otherwise failures print the password
+      property :sensitive, [TrueClass, FalseClass],
+               default: true
+
+      action :join do
+        description "Update the workgroup."
+
+        unless workgroup_member?
+          cmd = "$pswd = ConvertTo-SecureString \'#{new_resource.password}\' -AsPlainText -Force;" if new_resource.password
```
Comment (stuartpreston) [inline lib/chef/resource/windows_workgroup.rb]: As new_resource.password is conditional, `cmd` should be instantiated separately or the following lines will fail.

Comment (stuartpreston) [reply lib/chef/resource/windows_workgroup.rb]: i.e. 
```
cmd = ""
cmd << "whatever" if whatever
```

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -543,6 +543,10 @@ Use the windows_printer_port resource to create and delete TCP/IPv4 printer port
 
 Use the windows_shortcut resource to create shortcut files on Windows. This resource was ported from the `windows` community cookbook.
 
+### windows_workgroup
+
+Use the windows_workgroup resource to join a Windows Workgroup and reboot the node. This resource is based on the `windows_ad_join` resource.
```
Comment (stuartpreston) [inline RELEASE_NOTES.md]: We may choose to modify this slightly before release.

**Comment by stuartpreston** [comment]
Ok so I have done an initial test today for a non-domain joined machine, and plan to progress the rest of the tests after a couple of necessary fixes (marked in the review)

**Scenario: Windows 2016 machine (default Azure install) joining workgroup 'SPWORKGROUP'**
```
C:\projects\chef\chef>bundle exec chef-apply repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_workgroup[SPWORKGROUP] action join

    ================================================================================
    Error executing action `join` on resource 'windows_workgroup[SPWORKGROUP]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0], but received '1'
    ---- Begin output of powershell.exe -NoLogo -NonInteractive -NoProfile -ExecutionPolicy Unrestricted -InputFormat None -Command "Get-WmiObject -Class Win32_ComputerSystem).Workgroup" ----
    STDOUT:
    STDERR: At line:1 char:42
    + Get-WmiObject -Class Win32_ComputerSystem).Workgroup
    +                                          ~
    Unexpected token ')' in expression or statement.
```

Fixed typo in PowerShell command and tried again:

```
C:\projects\chef\chef>bundle exec chef-apply repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_workgroup[SPWORKGROUP] action join

    ================================================================================
    Error executing action `join` on resource 'windows_workgroup[SPWORKGROUP]'
    ================================================================================

    NoMethodError
    -------------
    undefined method `<<' for nil:NilClass

```
This is caused when no password is supplied in the recipe as the string that is built up is never instantiated in this scenario. (suggestion in review)

```
C:\projects\chef\chef>bundle exec chef-apply repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_workgroup[SPWORKGROUP] action join
    - join workstation workgroup SPWORKGROUP
    * reboot[Reboot to join workgroup SPWORKGROUP] action reboot_now[2018-09-13T15:39:41+00:00] WARN: Rebooting system immediately, requested by 'Reboot to join workgroup SPWORKGROUP'
[2018-09-13T15:39:41+00:00] WARN: Rebooting system immediately, requested by 'Reboot to join workgroup SPWORKGROUP'
```

Workgroup "join" was successful (in this scenario no reboot is required, I'm happy to keep the default reboot as :immediate as it can be overridden) 

*Idempotency check* - successful
```
C:\projects\chef\chef>bundle exec chef-apply repro.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * windows_workgroup[SPWORKGROUP] action join (up to date)
```

If you can fix those issues, I'll try and complete the domain join/unjoin tests tomorrow so we can get this into the next release.

File: lib/chef/resources.rb [hunk 6]
```diff
@@ -139,4 +140,4 @@
 require "chef/resource/windows_printer_port"
 require "chef/resource/windows_shortcut"
 require "chef/resource/windows_task"
-require "chef/resource/locale"
```
Comment (tas50) [inline lib/chef/resources.rb]: This part needs to go back in. I think this was removed during the rebase

Comment (stuartpreston) [reply lib/chef/resources.rb]: I moved it to line 60 (I hope)

Comment (tas50) [reply lib/chef/resources.rb]: Oops. Yep it's still there

**Pull request merged**